### PR TITLE
Add support for custom calendar days and fluid calendar sizing

### PR DESCRIPTION
**Pull request by cliffpyles**
Adds support for custom rendering of days. Also adds support for fluid sizing of the component. Issue #4481 

#### What testing has been done on this PR?

1. Manually tested new and existing Calendar stories.
2. Added two new snapshots. One for the children prop and one for the fill prop.

#### How should this be manually tested?

The custom day can be tested by navigating to the "Custom day" Calendar story. There is one implementation for the story with an onSelect prop and one implementation without the onSelect prop.

The fluid calendar sizing can be tested by navigating to the "Fill" Calendar story.

#### Screenshots (if appropriate)

**Custom Day**

image

**Fluid Sizing**

image

#### Do the grommet docs need to be updated?

Yes.

#### Should this PR be mentioned in the release notes?

I think so.

#### Is this change backwards compatible or is it a breaking change?

It is backwards compatible.

### CODE REVIEW

File: src/js/components/Calendar/Calendar.js [hunk 1]
```diff
@@ -82,15 +82,90 @@ const buildDisplayBounds = (reference, firstDayOfWeek) => {
...
 
+const CalendarDayButton = ({
+  a11yTitle,
+  active,
+  disabled,
+  fill,
+  onClick,
+  onMouseOver,
+  onMouseOut,
+  children,
```
Comment (ericsoderberghp) [inline src/js/components/Calendar/Calendar.js:93]: I think we can just do something like:
```
const CalendarDayButton = props => <Button tabIndex={-1} plain {...props} />;
```

File: src/js/components/Calendar/Calendar.js [hunk 2]
```diff
@@ -82,15 +82,90 @@ const buildDisplayBounds = (reference, firstDayOfWeek) => {
...
   (
     {
       animate = true,
       bounds: validBounds,
+      children,
       date: dateProp,
       dates: datesProp,
       daysOfWeek,
       disabled,
+      fill,
```
Comment (ericsoderberghp) [inline src/js/components/Calendar/Calendar.js:168]: We need to update Calendar/doc.js and Calendar/index.d.ts for `fill` and `children`.

File: src/js/components/Calendar/StyledCalendar.js [hunk 3]
```diff
@@ -61,35 +73,50 @@ const slideStyle = props => {
...
+const weeksSizeStyle = props => {
+  return css`
+    ${props.fillContainer && 'display: flex;'}
+    ${props.fillContainer && 'flex-direction: column;'}
+    ${props.fillContainer && 'height: 100%;'}
+  `;
+};
 const StyledWeeks = styled.div`
   position: relative;
+  ${props => weeksSizeStyle(props)}
```
Comment (ericsoderberghp) [inline src/js/components/Calendar/StyledCalendar.js:85]: Would be a bit simpler to:
`${props => props.fillContainer && weeksSizeStyle()`

File: src/js/components/Calendar/StyledCalendar.js [hunk 4]
```diff
@@ -61,35 +73,50 @@ const slideStyle = props => {
...
+    display: flex;
+    justify-content: space-between;
+    ${props.fillContainer && 'flex: 1;'}
+  `;
+};
 const StyledWeek = styled.div`
-  display: flex;
-  flex-direction: row;
-  flex-justify: between;
+  ${props => weekSizeStyle(props)}
```
Comment (ericsoderberghp) [inline src/js/components/Calendar/StyledCalendar.js:100]: Can we keep `display` and `justify-content` here and then: `${props => props.fillContainer && 'flex: 1;'}`?

File: src/js/components/Calendar/StyledCalendar.js [hunk 5]
```diff
@@ -61,35 +73,50 @@ const slideStyle = props => {
...
 `;
 
 StyledWeek.defaultProps = {};
 Object.setPrototypeOf(StyledWeek.defaultProps, defaultProps);
 
 const StyledDayContainer = styled.div`
   flex: 0 0 auto;
+  ${props => props.fillContainer && 'width: 14.3%;'}
```
Comment (ericsoderberghp) [inline src/js/components/Calendar/StyledCalendar.js:108]: Can we add a comment here about why "14.3%", that it is 100/7?

**Review by ericsoderberghp — COMMENTED** [review]
Great progress! Just a few comments.

File: src/js/components/Calendar/Calendar.js (hunk 2 above)
Comment (cliffpyles) [reply src/js/components/Calendar/Calendar.js:141]: Updates made.

File: src/js/components/Calendar/StyledCalendar.js (hunk 5 above)
Comment (cliffpyles) [reply src/js/components/Calendar/StyledCalendar.js:105]: Sounds good. Update made.

File: src/js/components/Calendar/Calendar.js (hunk 1 above)
Comment (cliffpyles) [reply src/js/components/Calendar/Calendar.js]: Nice. Much cleaner. Update made.

File: src/js/components/Calendar/StyledCalendar.js (hunk 3 above)
Comment (cliffpyles) [reply src/js/components/Calendar/StyledCalendar.js]: Sounds good. Update made.

File: src/js/components/Calendar/StyledCalendar.js (hunk 4 above)
Comment (cliffpyles) [reply src/js/components/Calendar/StyledCalendar.js]: Sounds good. Update made.

**Comment by cliffpyles** [comment]
Thanks! I made the updates requested. Anything else?

**Comment by ShimiSun** [comment]
Hi buddy, for the hash code failure, can you please try:
```
rm -rf node_modules/grommet-icons
yarn cache clean
yarn install
```
Push the updates, and then we can see whether it fixed the problem.

**Comment by cliffpyles** [comment]
Hi @ShimiSun. Thanks. I tried the removing the node_modules, cleaning the cache, and reinstalling, but unfortunately that doesn't resolve the issue for me. I still get an error complaining about the integrity check failing.

```
error https://github.com/grommet/grommet-icons/tarball/stable: Integrity check failed for "grommet-icons" (computed integrity doesn't match our records, got "sha512-llFsMjqkLhw4d90pvwQhJlzA8yRZqJM2WD91egjtwew2Ut83CEavLn1tUYRCPk7GsKlGzRuFYa+6PfmeTOdB5w== sha1-oGA8pOXcDBgIX2ifYvkJMMNvEII=")
```

I also tried removing the lock file completely, and that did allow me to install the dependencies again and run storybook successfully, but unfortunately the tests started blowing up after I did that. :( I'm guessing it installed the incorrect versions of some of the dependencies after I removed the lock?

I also tried cleaning the yarn cache globally, and updating my version of yarn. No luck with those. Any other ideas?

**Comment by ShimiSun** [comment]
>Hi @ShimiSun. Thanks. I tried the removing the node_modules, cleaning the cache, and reinstalling, but unfortunately that doesn't resolve the issue for me. I still get an error complaining about the integrity check failing.

You were right. I fixed the issue with #4528 and some dependency updates on the master branch and it should work now, I see your Netlify deployment is working as expected so you no longer have the hash issue.

> I have also tried removing the lock file completely,

This won't work, since `styled-components` recent update is breaking our snapshots, https://github.com/styled-components/styled-components/releases/tag/v5.2.0, I'll take a closer look into it tomorrow, just make sure you are not removing/creating the yarn.lock file, the build will surely break if you do.

Your current test failures should also be fixed now, so run `yarn test-update` while you make sure you are using node version 14.4 and higher and it should fix the snapshots and you can push them to the branch to fix the failing tests.

Thank you for your patience and collaboration! 💟

**Comment by cliffpyles** [comment]
That did the trick! Dependencies installed correctly and the tests are all passing for me. Thanks so much for your help!

**Review by ericsoderberghp — APPROVED** [review]

**Comment by ericsoderberghp** [comment]
Thanks for your great contribution!

**Comment by cliffpyles** [comment]
Glad to pitch in! Hopefully it will be one of many!

**Pull request merged**